src/Products/Viewer/Cache/ExternalResourcesStreamFactory.cs

```diff
@@ -0,0 +1,43 @@
+﻿using GroupDocs.Viewer.Interfaces;
+using GroupDocs.Viewer.Results;
+using System.IO;
+
+namespace GroupDocs.Viewer.MVC.Products.Viewer.Cache
+{
+    public class ExternalResourcesStreamFactory : IPageStreamFactory, IResourceStreamFactory
+    {
+        // Resources storage path.
+        private readonly string _outputPath;
+
+        // URL for resources download.
+        private readonly string _urlPrefix;
+
+        public ExternalResourcesStreamFactory(string cachePath, string resourcesPath)
+        {
+            if (!Directory.Exists(cachePath))
+            {
+                Directory.CreateDirectory(cachePath);
+            }
+
+            _outputPath = cachePath;
+            _urlPrefix = resourcesPath;
+        }
+
+        public Stream CreatePageStream(int pageNumber) =>
+            File.OpenWrite(Path.Combine(_outputPath, $"p_{pageNumber}.html"));
+
+        public void ReleasePageStream(int pageNumber, Stream pageStream) =>
```

Comment:
Codacy found an issue: [Make 'ReleasePageStream' a static method.](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5933705)